rust/lance/src/io/commit.rs

```diff
@@ -154,20 +154,27 @@ pub(crate) async fn commit_new_dataset(
 ///
 /// Fields such as `physical_rows` and `num_deleted_rows` may not have been
 /// in older datasets. To bring these old manifests up-to-date, we add them here.
-async fn migrate_manifest(dataset: &Dataset, manifest: &mut Manifest) -> Result<()> {
-    if manifest.fragments.iter().all(|f| {
-        f.physical_rows > 0
-            && (f.deletion_file.is_none()
-                || f.deletion_file
-                    .as_ref()
-                    .map(|d| d.num_deleted_rows)
-                    .unwrap_or_default()
-                    > 0)
-    }) {
+async fn migrate_manifest(
+    dataset: &Dataset,
+    manifest: &mut Manifest,
+    recompute_stats: bool,
+) -> Result<()> {
+    if !recompute_stats
+        && manifest.fragments.iter().all(|f| {
+            f.physical_rows > 0
+                && (f.deletion_file.is_none()
+                    || f.deletion_file
+                        .as_ref()
+                        .map(|d| d.num_deleted_rows)
+                        .unwrap_or_default()
+                        > 0)
+        })
+    {
         return Ok(());
     }
 
-    manifest.fragments = Arc::new(migrate_fragments(dataset, dataset.fragments()).await?);
+    manifest.fragments =
+        Arc::new(migrate_fragments(dataset, &manifest.fragments, recompute_stats).await?);
```

Comment:
I found another critical bug here 😞 

We migrate the old fragments not the new, so this basically is rolling back the transactions changes. This means data loss in most cases.

Reply:
Could we have some tests to cover this case?

Reply:
Working on adding tests now.

Reply:
Tests have been added.

Reply:
I'm confused right now. This seems like it would have caused new fragments to be ignored, but my repro in https://github.com/lancedb/lance/issues/1531 does not seem to cause this.

Reply:
Figured it out, the data loss bug was added in a later version.